src/lib/storage/dictionary_segment.cpp

```diff
@@ -1,24 +1,62 @@
-#include "dictionary_segment.hpp"
+#include <set>
+#include <unordered_map>
+#include <limits>
 
+#include "dictionary_segment.hpp"
+#include "fixed_width_integer_vector.hpp"
 #include "utils/assert.hpp"
+#include "types.hpp"
+#include "type_cast.hpp"
 
 namespace opossum {
 
 template <typename T>
 DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
-  // Implementation goes here
+
+  // determine unique values and store in sorted set
+  auto dict_values = std::set<T>{};
+  auto segment_size = abstract_segment->size();
+  for (auto value_index = ChunkOffset{0}; value_index < segment_size; ++value_index) {
+    dict_values.emplace(type_cast<T>((*abstract_segment)[value_index]));
+  }
+
+  // select the right attribute vector integer type
+  auto n_unique_values = dict_values.size();
+  Assert(n_unique_values <= std::numeric_limits<uint32_t>::max(), "Too many unique values");
+  if (n_unique_values <= std::numeric_limits<uint8_t>::max()) {
```

Comment:
<= does not leave space for INVALID_VALUE_ID (in this case == std::numeric_limits<uint8_t>::max())